### PR TITLE
fix(animations): Ensure elements are removed from the cache after leave animation.

### LINKED ISSUE #24197: Memory leak with animations in combination with @HostBinding
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x ] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
When using animations in combination with a @HostBinding dom nodes are not released.

Open https://angular-5-example-tm3ghr.stackblitz.io/bar in chrome and open the performance
monitor. Toggle on "DOM nodes". Note the starting number, now switch between foo and bar a number of times. Now click "Collect garbage" observe that the "DOM nodes" will keep growing.

## Expected behavior
That the DOM nodes are garbage collected, and that the DOM nodes do not increase indefinitly.

## Minimal reproduction of the problem with instructions
https://stackblitz.com/edit/angular-5-example-tm3ghr?file=app%2Fcontent-state.component.ts

## Environment

<pre><code>
Angular version: 5.2.10

Browser:
- [x ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX

### PR DESCRIPTION
**Pull request by JeanMeche**
This commit fixes a memory leak where references were retained in multiple `Map` structures. 

`_namespaceLookup` was cleared before the call to `processLeaveNode()` which was using the lookup.
Without that lookup `clearElementCache()` wasn't called thus keeping a reference to the element.  

Fixes #24197 & #50533

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix

## Does this PR introduce a breaking change?

- [x] No

### CODE REVIEW

File: packages/animations/browser/test/render/transition_animation_engine_spec.ts [hunk 1]
```diff
@@ -113,13 +113,20 @@ describe('TransitionAnimationEngine', () => {
...
+          .toBe(true, 'Expected element data to be defined.');
 
       expect(engine.elementContainsData(DEFAULT_NAMESPACE_ID, element)).toBeTruthy();
 
+      engine.destroy(DEFAULT_NAMESPACE_ID, null);
       engine.removeNode(DEFAULT_NAMESPACE_ID, element, true);
       engine.flush();
+      engine.players[0].finish();
 
-      expect(engine.elementContainsData(DEFAULT_NAMESPACE_ID, element)).toBeTruthy();
```
Comment (JoostK) [inline packages/animations/browser/test/render/transition_animation_engine_spec.ts:122]: This test definitely looked broken before, with both these `elementContainsData` calls being checked to return true even after removal and flush. I'd suggest also removing the first `elementContainsData` assertion as it seems rather meaningless with the new assertions, and it allows dropping test-only `elementContainsData` implementations from production code (both from `TransitionAnimationsEngine` and `AnimationTransitionNamespace`).

File: packages/animations/browser/src/render/transition_animation_engine.ts [hunk 2]
```diff
@@ -645,14 +645,16 @@ export class TransitionAnimationEngine {
...
       const index = this._namespaceList.indexOf(ns);
       if (index >= 0) {
         this._namespaceList.splice(index, 1);
       }
     });
 
-    this.afterFlushAnimationsDone(() => ns.destroy(context));
+    this.afterFlushAnimationsDone(() => {
+      ns.destroy(context);
+      delete this._namespaceLookup[namespaceId];
```
Comment (JoostK) [inline packages/animations/browser/src/render/transition_animation_engine.ts:656]: It's a bit unfortunate to me that cleanup logic is (still) spread across two phases, especially now that we've seen that first-phase cleanup can remove needed state too early. Apparently the destroyed namespace is still interacting with this engine, so arguably we should not be cleaning it up until those interactions can no longer occur entirely?

tldr; can all of the `afterFlush` work be pushed into this second phase, or would that result in issues?

**Review by JoostK — COMMENTED** [review]
Thanks for investigating, this must have been hard to track down!

File: packages/animations/browser/src/render/transition_animation_engine.ts (hunk 2 above)
Comment (JeanMeche) [reply packages/animations/browser/src/render/transition_animation_engine.ts:656]: Looks like, we're good with moving everything into `afterFlushAnimationsDone` !

File: packages/animations/browser/test/render/transition_animation_engine_spec.ts (hunk 1 above)
Comment (JeanMeche) [reply packages/animations/browser/test/render/transition_animation_engine_spec.ts:122]: Great suggestion ! I'll remove it !

**Review by jessicajaniuk — APPROVED** [review]
Nice work fixing this!

**Comment by dylhunn** [comment]
This PR was merged into the repository by commit a14bdfe8591a33d359bf4940f4efa828499a6373.
